### DISCUSSION TITLE
Mapping CTRL+C in both Wezterm and NVIM

### DISCUSSION THREAD

**Discussion by cmvanb** [wezterm/wezterm#2426] [Q&A]
Hello,

I'm trying to get consistent copy key bindings across my terminal (Wezterm), text editor (NVIM) and browser (Firefox). I am aware of the legacy of terminal escape codes, but I don't see any technical reason this shouldn't work. Alas, I'm unable to make it work for both NVIM and Wezterm.

wezterm.lua
```
{ key = 'c', mods = 'CTRL', action = wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' } },
```

init.vim
```
vnoremap <C-c> "+y
```

- If I assign the binding for Wezterm, it does not work in NVIM.

- If I assign the binding for NVIM, it does not work in Wezterm (obviously).

- If I assign bindings for both, it does not work in NVIM.

I believe Wezterm's binding is overriding/intercepting the NVIM binding. Is there a way to pass the keypresses to NVIM instead of intercepting them? Or is there a workaround to allow Wezterm to paste in NVIM?

Hoping for some insight! Really love Wezterm, best terminal I've come across.

**Comment by wez** [comment]
Yes, your wezterm ctrl-c binding prevents the default behavior of sending CTRL-C to the application.

What you can do instead is:

```lua
action = wezterm.action_callback(function(window, pane)
   if pane:is_alt_screen_active() then
     -- allow "full screen" TUI apps to receive and handle CTRL-C for themselves
     window:perform_action(wezterm.action.SendKey{key='c', mods='CTRL'})
   else
     -- otherwise, treat it as a copy operation
     window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' })
   end
end)
```

Alternatively, you could use [pane:get_foreground_process_name()](https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html) to test for vim/nvim explicitly instead of matching on any full-screen app.

However: note that CTRL-C for non-TUI apps is generally configured via `stty` to be interpreted as SIGINT when it is processed by the kernel, so **this particular key assignment will prevent you from using CTRL-C to interrupt a program**.  You can use `stty` to redefine the `intr` setting to a different key combination if you wish, or define a different key assignment in wezterm that will `SendKey{key='c', mods='CTRL'}` for the situations where you need it.

* https://wezfurlong.org/wezterm/config/lua/wezterm/action_callback.html
* https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html
* https://wezfurlong.org/wezterm/config/lua/pane/is_alt_screen_active.html

**Reply by MuhammedZakir** [reply]
@*cmvanb*: If you use a terminal/shell that is not configured for this, you could easily kill the program you're running when you try to copy via  `Ctrl`+`c`. Depending on the programs you use, this can be disastrous. I would deprecate remapping "problematic" keys such as `Ctrl`+`c` and `Ctrl`+`z` because of potential, possibly disastrous, consequences. But, it's upto you! :-)

**Reply by cmvanb** [reply]
> Yes, your wezterm ctrl-c binding prevents the default behavior of sending CTRL-C to the application.
> 
> What you can do instead is:
> 
> ```lua
> action = wezterm.action_callback(function(window, pane)
>    if pane:is_alt_screen_active() then
>      -- allow "full screen" TUI apps to receive and handle CTRL-C for themselves
>      window:perform_action(wezterm.action.SendKey{key='c', mods='CTRL'})
>    else
>      -- otherwise, treat it as a copy operation
>      window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' })
>    end
> end)
> ```
> 
> Alternatively, you could use [pane:get_foreground_process_name()](https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html) to test for vim/nvim explicitly instead of matching on any full-screen app.
> 
> However: note that CTRL-C for non-TUI apps is generally configured via `stty` to be interpreted as SIGINT when it is processed by the kernel, so **this particular key assignment will prevent you from using CTRL-C to interrupt a program**. You can use `stty` to redefine the `intr` setting to a different key combination if you wish, or define a different key assignment in wezterm that will `SendKey{key='c', mods='CTRL'}` for the situations where you need it.
> 
> * https://wezfurlong.org/wezterm/config/lua/wezterm/action_callback.html
> * https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html
> * https://wezfurlong.org/wezterm/config/lua/pane/is_alt_screen_active.html

Thank you for the quick and detailed answer!

I tried your suggestion, but it's not working for me. With the following bindings in `wezterm.lua`:

```
keys = {
    {
        key = 'c',
        mods = 'CTRL',
        action = wezterm.action_callback(function(window, pane)
            if pane:is_alt_screen_active() then
                window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' })
                wezterm.log_info('alt screen active')
            else
                window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' })
                wezterm.log_info('regular copy')
            end
        end),
    },
    { key = 'v', mods = 'CTRL', action = wezterm.action{ PasteFrom = 'Clipboard' } },
},
```

I not able to copy regular terminal output in any Wezterm instance, or while running NVIM or in any other program...

Based on the documentation, I expected to see the log messages in the parent terminal, but I do not. I do see one log message from Wezterm (`wezterm_gui > Spawned your command via the existing GUI instance. ...`), but not my own messages. Do I need to enable a logging flag or something? Or is it simply not reaching that log line?

I tried the alternative solution with `get_foreground_process_name`, also no luck.

```
keys = {
    {
        key = 'c',
        mods = 'CTRL',
        action = wezterm.action_callback(function(window, pane)
            if pane:get_foreground_process_name() == 'nvim' then
                window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' })
                wezterm.log_info('nvim active')
            else
                window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' })
                wezterm.log_info('regular copy')
            end
        end),
    },
    { key = 'v', mods = 'CTRL', action = wezterm.action{ PasteFrom = 'Clipboard' } },
},
```

Can you see any mistake in my bindings?

edit:
I should note, I tried the following simpler binding on a different key combination to debug- it also does not work:

```
{
    key = 'x',
    mods = 'CTRL',
    action = wezterm.action_callback(function(window, pane)
        window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' })
    end),
    -- action = wezterm.action({ CopyTo = 'ClipboardAndPrimarySelection' })
},
```

If I switch back to the wezterm.action syntax, it works.

**Reply by MuhammedZakir** [reply]
`window:perform_action` needs one more arg, `pane`. E.g.
```lua
window:perform_action(wezterm.action.CopyTo 'ClipboardAndPrimarySelection', pane)
```

**Reply by cmvanb** [reply]
Thank you @MuhammedZakir ! That did the trick. For future reference, here is the complete CTRL+C binding that passes the keypress to NVIM and other 'alt screen' programs:

```
{
    key = 'c',
    mods = 'CTRL',
    action = wezterm.action_callback(function(window, pane)
        if pane:is_alt_screen_active() then
            window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' }, pane)
        else
            window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' }, pane)
        end
    end),
},
```

Keeping in mind that it overrides CTRL+C behavior in all programs, now I'm going to rebind the stty interrupt signal :smiling_imp: 

Thanks to both for your quick assistance! Much appreciated. Wezterm rocks!

**Comment by pmarreck** [comment]
For what it's worth, I came up with a similar config to make wezterm behave like kitty with regards to control-C. If there is text selected, it will copy it, otherwise it will pass the interrupt through. Not sure how to handle control-V yet though haha
```
  keys = {
    {
      key = 'c',
      mods = 'CTRL',
      action = wezterm.action_callback(function(window, pane)
        local sel = window:get_selection_text_for_pane(pane)
        if (not sel or sel == '') then
          window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' }, pane)
        else
          window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' }, pane)
        end
      end),
    },
  },
  ```

**Reply by pmarreck** [reply]
I went further. ctrl-V now pastes; ctrl-shift-V sends ctrl-V to the shell to do its "verbatim" input mode.
alt-C/V copies/pastes. ctrl-C behaves as described above. Adding here simply to help out anyone who googled their way here (as I just did):
```
  keys = {
    {
      key = 'c',
      mods = 'CTRL',
      action = wezterm.action_callback(function(window, pane)
        local sel = window:get_selection_text_for_pane(pane)
        if (not sel or sel == '') then
          window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' }, pane)
        else
          window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' }, pane)
        end
      end),
    },
    { key = 'v', mods = 'CTRL', action = wezterm.action.Paste },
    { key = 'v', mods = 'SHIFT|CTRL', action = wezterm.action_callback(function(window, pane)
      window:perform_action(wezterm.action.SendKey{ key='v', mods='CTRL' }, pane) end),
    },
    { key = 'V', mods = 'SHIFT|CTRL', action = wezterm.action_callback(function(window, pane)
      window:perform_action(wezterm.action.SendKey{ key='v', mods='CTRL' }, pane) end),
    },
    { key = 'c', mods = 'ALT', action = wezterm.action.Copy },
    { key = 'v', mods = 'ALT', action = wezterm.action.Paste },
  },
  ```

**Reply by laravieira** [reply]
Just to follow up, some functions on that config are deprecated (`Paste` and `Copy`), so here is what worked for me on version `20240203-110809-5046fc22` on the `~/.wezterm.lua` file:
```lua
config.keys = {
    {
      key = 'c',
      mods = 'CTRL',
      action = wezterm.action_callback(function(window, pane)
        local sel = window:get_selection_text_for_pane(pane)
        if (not sel or sel == '') then
          window:perform_action(wezterm.action.SendKey{ key='c', mods='CTRL' }, pane)
        else
          window:perform_action(wezterm.action{ CopyTo = 'ClipboardAndPrimarySelection' }, pane)
        end
      end),
    },
    { key = 'v', mods = 'CTRL', action = wezterm.action.PasteFrom 'Clipboard' },
    { key = 'v', mods = 'SHIFT|CTRL', action = wezterm.action_callback(function(window, pane)
      window:perform_action(wezterm.action.SendKey{ key='v', mods='CTRL' }, pane) end),
    },
    { key = 'V', mods = 'SHIFT|CTRL', action = wezterm.action_callback(function(window, pane)
      window:perform_action(wezterm.action.SendKey{ key='v', mods='CTRL' }, pane) end),
    },
    { key = 'c', mods = 'ALT', action = wezterm.action.CopyTo 'ClipboardAndPrimarySelection' },
    { key = 'v', mods = 'ALT', action = wezterm.action.PasteFrom 'Clipboard' },
}
```
